smart_open/smart_open_lib.py

```diff
@@ -617,16 +617,21 @@ def compression_wrapper(file_obj, filename, mode):
     If the filename extension isn't recognized, will simply return the original
     file_obj.
     """
-    _, ext = os.path.splitext(filename)
+    root, ext = os.path.splitext(filename)
+    
     if ext == '.bz2':
         if IS_PY2:
             from bz2file import BZ2File
         else:
             from bz2 import BZ2File
+        if 'http' in root:
+            return make_closing(HttpReadStream)(filename, mode)
         return make_closing(BZ2File)(filename, mode)
 
     elif ext == '.gz':
         from gzip import GzipFile
+        if 'http' in root:
+            return make_closing(HttpReadStream)(filename, mode)
         return make_closing(GzipFile)(filename, mode)
```

Comment:
This doesn't fix the resource leak -- it just ignores the compression extension and then re-opens the URL, downloading it a second time. That's not correct.